### ISSUE TITLE
Truncated http requests

### ISSUE THREAD

**Issue by J-A-M** [django/daphne#94]
Hi,

I'm having an urgent problem with Daphne when it's handling large http requests.
A large javascript module (phaser.js) is being truncated.  Usually at one of two points.  There is no error (other than the a javascript syntax error).
This only happens when the server is a VM guest and the client is the VM host.
When the server and client are the same (vm guest) it does not happen.
I've tried with in-memory, ipc and redis backends, all behave the same. So I think I can rule out the connection between Daphne and the worker, it's more like the handling of TCP by Daphne.

As a fix for the is urgently required, obviously I can serve http using a separate server or wsgi as a work-round. However, I would like to get a proper fix.  What might be the issue.  What debugging can I do?

**Comment by andrewgodwin** [comment]
I need steps to replicate before I can fix the problem; given your steps to replicate, I suspect it might be the virtual machine driver you are using instead (e.g. VirtualBox networking has known issues with sendfile()). What virtual machine solution and version are you using? How big is the file that you're sending? Where does it get cut off in terms of bytes?

**Comment by J-A-M** [comment]
No worries, I think I've resolved it.  I was about to cancel the report. Thanks for the prompt interest anyway.   It was a strange one to do with caching and how Daphne must behave when denying requests.

IP Address of virtual adapter was not in ALLOWED_HOSTS.  Django was not complaining about that.

The client (I tried Firefox/IE/Edge/Chrome - all the same) had received half of the file and then cached it. Daphne must somehow half serve the file.  Server was responding 304, so it was using the (broken) cached version (Phaser is 3.5Mb).

Adding correct IP to ALLOWED_HOSTS fixed it.
